File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_matlab_labs_and_solutions/observer/solution/DpMap.m
% DpMap.m

classdef DpMap < handle
	properties (GetAccess = private, SetAccess = private)
		List % DpList of DpPair
	end

	methods
		% constructor:
		function theDpMap = DpMap
			theDpMap.List = DpList;
		end

		function put(theDpMap, aKey, aValue)
			theDpMap.List.add(DpPair(aKey, aValue));
		end

		function result = contains(theDpMap, aKey)
			aIterator = theDpMap.List.makeIterator;
			while ~aIterator.isDone
				aPair = aIterator.getCurrent;
				if isa(aPair.First, 'char') && isa(aKey, 'char') && strcmp(aPair.First, aKey)
					result = true;
					return;
				elseif ~isa(aPair.First, 'char') && ~isa(aKey, 'char') && aPair.First == aKey
					result = true;
					return;
				end
				aIterator.advance;
			end
			result = false;
		end

		function result = get(theDpMap, aKey)
			aIterator = theDpMap.List.makeIterator;
			while ~aIterator.isDone
				aPair = aIterator.getCurrent;
				if isa(aPair.First, 'char') && isa(aKey, 'char') && strcmp(aPair.First, aKey)
					result = aPair.Second;
					return;
				elseif ~isa(aPair.First, 'char') && ~isa(aKey, 'char') && aPair.First == aKey
					result = aPair.Second;
					return;
				end
				aIterator.advance;
			end
			result = {}; % value_type.empty;
		end
	end
end

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/builder/lab/DpMap.java
// DpMap.java

public class DpMap<Key, Value>
{
	public DpMap()
	{
		myList = new DpList<DpPair<Key, Value> >();
	}

	void put(Key aKey, Value aValue)
	{
		myList.add(new DpPair<Key, Value>(aKey, aValue));
	}

	public boolean contains(Key aKey)
	{
		DpIterator<DpPair<Key, Value> > aIterator = myList.makeIterator();
		while (!aIterator.isDone())
		{
			DpPair<Key, Value> aPair = aIterator.getCurrent();
			if (aPair.getFirst() == aKey)
			{
				return true;
			}
			aIterator.advance();
		}
		return false;
	}

	public Value get(Key aKey)
	{
		DpIterator<DpPair<Key, Value> > aIterator = myList.makeIterator();
		while (!aIterator.isDone())
		{
			DpPair<Key, Value> aPair = aIterator.getCurrent();
			if (aPair.getFirst() == aKey)
			{
				return aPair.getSecond();
			}
			aIterator.advance();
		}
		throw new DpException("DpMap<T>.get() - aKey is not in this DpMap.");
	}

	private DpList<DpPair<Key, Value> > myList;
}
